### PR TITLE
CORE-18088 Race condition registering RPC endpoints in JavalinServer

### PR DESCRIPTION
**Pull request by mbrkic-r3**
Synchronized registration/removal of endpoints in `JavalinServer`.

### CODE REVIEW

**Review by vkolomeyko — DISMISSED** [review]
LGTM

File: libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt [hunk 1]
```diff
@@ -111,31 +111,35 @@ class JavalinServer(
...
-        if(null != server) registerEndpointInternal(endpoint)
-        // record the path in case we need to register when it's already started
-        endpoints.add(endpoint)
+        serverLock.withLock {
+            if (endpoints.any { it.path == endpoint.path && it.methodType == endpoint.methodType })
+                throw IllegalArgumentException("Endpoint with path ${endpoint.path} and method ${endpoint.methodType} already exists.")
+            // register immediately when the server has been started
+            if (null != server) registerEndpointInternal(endpoint)
+            // record the path in case we need to register when it's already started
+            endpoints.add(endpoint)
```
Comment (driessamyn) [inline libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:127]: If we add the lock around this, then I don't think `endpoints` need to be a concurrent hashmap anymore?

File: libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt [hunk 2]
```diff
@@ -111,31 +111,35 @@ class JavalinServer(
...
     override fun removeEndpoint(endpoint: Endpoint) {
-        endpoints.remove(endpoint)
-        // NOTE:
-        //  The server needs to be restarted to un-register the path. However, this means everything dependent on
-        //  this is impacted by a restart, which doesn't feel quite right.
-        //  This also means we can't really DOWN/UP the lifecycle status of this because this would end up in a
-        //  relentless yoyo-ing of this component as dependent components keep calling this function.
-        // TODO - review if it is really needed to de-register a path when a Subscription goes down, for example.
-        if(null != server) restartServer()
+        serverLock.withLock {
```
Comment (driessamyn) [inline libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:132]: Does this not risk a deadlock given there is a lock around restartServer?

**Review by driessamyn — COMMENTED** [review]
I think instead of this change, we just need an additional in `registerEndpointInternal` as `endpoints` is already thread safe, so the only thing missed out in the original PR, I think, is the lock for registering endpoints in a running server.

I think the current change risks a deadlock.

File: libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt (hunk 1 above)
Comment (mbrkic-r3) [reply libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:127]: yes, I've changed it

File: libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt (hunk 2 above)
Comment (mbrkic-r3) [reply libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:132]: Not sure how we could have a deadlock?

Comment (driessamyn) [reply libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:132]: There is a lock around `restartServer` which is called inside here.
I thought it was called in a second place, which could have resulted in a deadlock, but it's not so it _should_ be ok.
However, I still think it's risky as somebody may make a chance that calls it from another place, and there isn't a reason for the double lock.
I think maybe the lock in restart can be removed.

Comment (mbrkic-r3) [reply libs/web/web-impl/src/main/kotlin/net/corda/web/server/JavalinServer.kt:132]: There is only one lock and it's re-entrant.

**Review by driessamyn — APPROVED** [review]
LGTM

**Pull request merged**